src/similarity/similarity.py

```diff
@@ -0,0 +1,229 @@
+from database.song_segment import SongSegment
+
+MATCHES = 10
+BUCKET_SIZE = 60 * 100
+
+
+def flatten(l): return [item for sublist in l for item in sublist]
+
+
+def process_segment(segment, sr):
+    harmonic, percussive = librosa.effects.hpss(segment)
+
+    tempogram = librosa.feature.tempogram(
+        percussive, hop_length=1024, win_length=16)
+
+    mfcc = librosa.feature.mfcc(segment, hop_length=1024)
+
+    chromagram = librosa.feature.chroma_cqt(y=harmonic, sr=sr, hop_length=1024)
+
+    return mfcc, chromagram, tempogram
+
+
+def load_songs(songs):
+    seg_db = SongSegment()
+    segments = []
+    for song in songs:
+        song_id, filename = song
+        segments.append(_load_song(song_id, filename, seg_db, False))
+    seg_db.close()
+    return flatten(segments)
+
+
+def _load_song(song_id, filename, segments, force):
+    segs = segments.get_all_with_id(song_id)
+
+    segment_data = []
+
+    if (force or segs == []):
+        # No segments in db, which means no features in db
+        y, sr = librosa.load(filename)
+
+        for i in range(0, y.shape[0]//sr//5 - 1):
+
+            sample = y[(i * sr * 5):
+                       ((i + 1) * sr * 5)]
+
+            mfcc, chromagram, tempogram = process_segment(sample, sr)
+
+            _id = segments.add(song_id, i*5*1000, (i+1)*5*1000, mfcc.tobytes(),
+                               chromagram.tobytes(), tempogram.tobytes(), [])
+
+            feature = create_feature(mfcc, chromagram, tempogram)
+
+            segment_data.append((_id, song_id, i*5, feature))
+
+    else:
+        # There are segments in db, look for features
+        for i in range(0, len(segs)):
+            segment = segs[i]
+
+            feature = create_feature(np.frombuffer(segment['mfcc']), np.frombuffer(
+                segment['chroma']), np.frombuffer(segment['tempogram']))
+
+            segment_data.append((segment['_id'], song_id, i*5, feature))
+
+    return segment_data
+
+
+def process_db_segment(segment):
+    feature = create_feature(np.frombuffer(segment['mfcc']), np.frombuffer(
+        segment['chroma']), np.frombuffer(segment['tempogram']))
+
+    return (segment['_id'], segment['song_id'], segment['time_from'] // 1000, feature)
+
+
+def create_feature(mfcc, chroma, tempogram):
+    fm = mfcc
+    fc = chroma
+    ft = tempogram
+    if (fm.ndim != 1):
+        fm = fm.flatten()
+    if (fc.ndim != 1):
+        fc = fc.flatten()
+    if (ft.ndim != 1):
+        ft = ft.flatten()
+    return np.concatenate((fm, fc * 133, ft * 280))
+
+
+def create_bucket(segments):
+    params_cp = falconn.LSHConstructionParameters()
+    params_cp.dimension = len(segments[0])
+    params_cp.lsh_family = falconn.LSHFamily.CrossPolytope
+    params_cp.distance_function = falconn.DistanceFunction.EuclideanSquared
+    params_cp.l = 25
+    params_cp.num_rotations = 2
+    params_cp.seed = 5721840
+    params_cp.num_setup_threads = 0
```

Comment:
Should anything of the above be added to config?

Reply:
Probably not